### ISSUE TITLE
Beacon sending lesson data on close blocked by EasyPrivacy to uBO Trusted Site

### ISSUE THREAD

**Issue by PinkDuck** [uBlockOrigin/uBlock-issues#2453]
### Prerequisites

- [X] This is not a support issue or a question. For support, questions, or help, visit [/r/uBlockOrigin](https://www.reddit.com/r/uBlockOrigin/).
- [X] I performed a [cursory search of the issue tracker](https://github.com/uBlockOrigin/uAssets/issues?q=is%3Aissue) to avoid opening a duplicate issue.
- [X] I am running the latest version of uBlock Origin (uBO).

### I tried to reproduce the issue when...

- [X] uBO is the only extension.
- [X] uBO uses default lists and settings.
- [X] using a new, unmodified browser profile.

### Description

Student lesson data in a commercial system is being blocked by uBO when navigating away from an active lesson. The system first attempts a fetch() request, which fails in NS_BINDING_FAILURE then falls back to sendBeacon mechanism using text/plain JSON content. However, the async delay in sending beacon means the host can be a non-Trusted Site by the time it gets intercepted by uBO. So either the EasyPrivacy list needs an exception for this particular site, or uBO could be adjusted to allow ping requests sent to Trusted Sites before processing privacy lists.

Example blocked item extract from uBO Logger:

```
Filter	*$ping,3p
Filter list	EasyPrivacy
 
Context	www.edp24.co.uk
Partyness	(3) edp24.co.uk ⇒ ljcreatelms.com
Type	beacon
URL	https://www.ljcreatelms.com/ScormData.aspx/Terminate

@@||www.ljcreatelms.com/ScormData.aspx/Terminate$ping

Where www.ljcreatelms.com is a Trusted Site
```

### URL(s) where the issue occurs.

https://www.ljcreatelms.com (uBlock disabled) sendBeacon to https://www.edp24.co.uk/news (uBlock enabled)

Also https://www.ljcreatelms.co.uk/ and https://lp1.ljcreatelms.com/

### Screenshot(s)

![image](https://user-images.githubusercontent.com/155881/212357040-9bc15dd8-7ca8-4921-8dd9-20ed5dc3dab1.png)

### uBO version

1.46.0

### Browser name and version

Mozilla Firefox x64 108.0.2

### Settings

Added "www.ljcreatelms.com" (without quotes) to Trusted Sites, confirmed uBO inactive at that site

### Notes

Tried disabling uBO in Firefox Extensions, issue did not occur
Tried navigating to another trusted site, issue did not occur
Only navigating from trusted to untrusted blocks the Beacon being sent back to Trusted

**Comment by gwarser** [comment]
This works also in the other way around.

1. Whitelist `https://example.com/`
2. search ddg for it `https://duckduckgo.com/?q=https%3A%2F%2Fexample.com%2F&ia=web`
3. there will be a direct link to `example.com` in wiki widget on the right
4. click on it
5. uBO does not block ddg `beacon` neither by `||duckduckgo.com/t/` or `*$ping,3p` because it's called from whitelisted `example.com` context.


See also: https://github.com/uBlockOrigin/uBlock-issues/issues/2363

**Comment by gwarser** [comment]
Still not blocked when navigating to whitelisted page.

Still blocked when navigating from whitelisted page.

Trying with:

```
||duckduckgo.com/t/$badfilter
*$ping,3p,badfilter
*$ping,strict3p
```

Because looks like `*$ping,3p` is correctly not applied now. On `example.com` beacon has `duckduckgo.com` context, so is `1p`.

Similarly, with `*$ping,1p`.

**Comment by PinkDuck** [comment]
A key issue can be seen at [CanIUse/keepalive](https://caniuse.com/?search=keepalive) in that some user agents (Firefox included) claim Fetch API full support, but actually a Fetch during beforeunload/unload does not have guaranteed network delivery and can be seen failing with NS_BINDING_ABORT. So I added user agent detection to compensate by using Beacon API directly, the first of these which does not get blocked, though the second ping does oddly.

**Comment by gorhill** [comment]
> Still not blocked when navigating to whitelisted page.

I am aware, this is another issue in a different code path, not as easily fixed.
